src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt

```diff
@@ -105,103 +109,116 @@ private constructor(
-        }
+        handleChatError(lastMessage.error)
       } else {
         RateLimitStateManager.invalidateForChat(project)
         if (extensionMessage.messages?.isNotEmpty() == true &&
             extensionMessage.isMessageInProgress == false) {
           getCancellationToken().dispose()
         } else {
-
           if (extensionMessage.chatID != null && extensionMessage.messages != null) {
-            messages.zip(extensionMessage.messages).forEachIndexed { index, messagesPair ->
-              val (sessionMessage, responseMessage) = messagesPair
-              if (sessionMessage != responseMessage) {
+            extensionMessage.messages.forEachIndexed { index, incomingMessage ->
+              val sessionMessage = messages.getOrNull(index)
+              if (sessionMessage != incomingMessage) {
                 ApplicationManager.getApplication().invokeLater {
-                  addMessageAtIndex(responseMessage, index)
+                  addMessageAtIndex(incomingMessage, index)
                 }
               }
             }
           }
         }
       }
-    } catch (error: Exception) {
-      getCancellationToken().abort()
-      logger.error(CodyBundle.getString("chat-session.error-title"), error)
+    } catch (exception: Exception) {
+      handleException(exception)
     }
   }
 
-  private fun addErrorMessageAsAssistantMessage(stringMessage: String, index: Int) {
+  private fun handleChatError(chatError: ChatError) {
+    getCancellationToken().abort()
+    CodyAgentService.setAgentError(project, chatError.message)
+
+    val rateLimitError = chatError.toRateLimitError()
+    val errorMessage =
+        if (rateLimitError != null) {
+          RateLimitStateManager.reportForChat(project, rateLimitError)
+          when {
+            rateLimitError.upgradeIsAvailable ->
+                CodyBundle.getString("chat.rate-limit-error.upgrade")
+            else -> CodyBundle.getString("chat.rate-limit-error.explain")
+          }
+        } else CodyBundle.getString("chat.general-error").fmt(chatError.message, "")
+
+    addErrorMessageAsAssistantMessage(errorMessage)
+  }
+
+  private fun handleException(e: Exception) {
+    getCancellationToken().abort()
+    CodyAgentService.setAgentError(project, e)
+
+    val message = ((e.cause as? CodyAgentException) ?: e).message ?: e.toString()
+    val errorReportLink = CodyErrorSubmitter().getEncodedUrl(e.getThrowableText(), message)
+    val errorReportMsg = CodyBundle.getString("chat.general-error.report").fmt(errorReportLink)
+    addErrorMessageAsAssistantMessage(
+        CodyBundle.getString("chat.general-error").fmt(message, errorReportMsg))
+  }
+
+  private fun addErrorMessageAsAssistantMessage(chatMessage: String) {
     UIUtil.invokeLaterIfNeeded {
-      addMessageAtIndex(ChatMessage(Speaker.ASSISTANT, stringMessage), index)
+      addMessageAtIndex(ChatMessage(Speaker.ASSISTANT, chatMessage), messages.count() - 1)
```

Comment:
hmmm... 🤔 `messages.count() - 1` can change... I wonder if we are aware of it's modifications and threading. previously the index was calculated in the pooled thread. Now, it's edt